### PR TITLE
fix(github-runners): add deprecation flag

### LINKED ISSUE #27290: Mark unsupported and deprecated versions as `isDeprecated`
### Describe the proposed change(s).

When reviewing my PR to add the macOS 12 and 13 large runner variants, @viceice said: [^1]

> we should add `isDeprecated` to all known unsupported and deprecated versions.
> 
> https://github.com/renovatebot/renovate/blob/e6d088abb0cdf8a229d8f7e4b90b759670320f10/lib/modules/datasource/types.ts#L57

I don't see any `isDeprecated` references in the [`lib/modules/datasource/github-runners/index.ts` file](https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/github-runners/index.ts) yet. So this is probably something we still need to do.

Relevant snippets from the `actions/runner-images` readme: [^2]

> ## Available Images
> 
> | Image | YAML Label | Included Software |
> | --------------------|---------------------|--------------------|
> | Ubuntu 22.04 | `ubuntu-latest` or `ubuntu-22.04` | [ubuntu-22.04] |
> | Ubuntu 20.04 | `ubuntu-20.04` | [ubuntu-20.04] |
> | macOS 14 [beta] | `macos-14-large`| [macOS-14] |
> | macOS 14 Arm64 [beta] | `macos-14` or `macos-14-xlarge`| [macOS-14-arm64] |
> | macOS 13 | `macos-13` or `macos-13-large` | [macOS-13] |
> | macOS 13 Arm64 | `macos-13-xlarge` | [macOS-13-arm64] |
> | macOS 12 | `macos-latest`,`macos-12`, `macos-latest-large`, or `macos-12-large`| [macOS-12] |
> | macOS 11 [deprecated] | `macos-11`| [macOS-11] |
> | Windows Server 2022 | `windows-latest` or `windows-2022` | [windows-2022] |
> | Windows Server 2019 | `windows-2019` | [windows-2019] |

> ## Software and Image Support
> 
> ### Support Policy
> 
> - Tools and versions will typically be removed 6 months after they are deprecated or have reached end-of-life
> - We support (at maximum) 2 GA images and 1 beta image at a time. We begin the deprecation process of the oldest image label once the newest OS image label has been released to GA.
> - The images generally contain the latest versions of packages installed except for Ubuntu LTS where we mostly rely on the Canonical-provided repositories.
> 
> - Popular tools can have several versions installed side-by-side with the following strategy:
> 
> | Tool name | Installation strategy |
> |-----------|-----------------------|
> | Docker images | not more than 3 latest LTS OS\tool versions. New images or new versions of current images are added using the standard tool request process |
> | Java      | all LTS versions |
> | Node.js   | 3 latest LTS versions |
> | Go        | 3 latest minor versions |
> | Python <br/> Ruby | 5 most popular `major.minor` versions |
> | PyPy      | 3 most popular `major.minor` versions |
> | .NET Core | 2 latest LTS versions and 1 latest version. For each feature version only latest patch is installed |
> | GCC <br/> GNU Fortran <br/> Clang <br/> GNU C++ | 3 latest major versions |
> | Android NDK | 1 latest non-LTS, 2 latest LTS versions |
> | Xcode     | - all OS compatible versions side-by-side <br/> - for beta, GM versions - latest beta only <br/> - old patch versions are deprecated in 3 months |

> ### Image Deprecation Policy
> 
> - Images begin the deprecation process of the oldest image label once a new GA OS version has been released.
> - Deprecation process begins with an announcement that sets a date for deprecation
> - As it gets closer to the date, GitHub begins doing scheduled brownouts of the image
> - During this time there will be an Announcement pinned in the repo to remind users of the deprecation.
> - Finally GitHub will deprecate the image and it will no longer be available

[^1]: [@viceice's comment on my PR](https://github.com/renovatebot/renovate/pull/27229#pullrequestreview-1876308795)
[^2]: [`actions/runner-images` readme](https://github.com/actions/runner-images/blob/main/README.md)

### PR DESCRIPTION
**Pull request by viceice**
## Changes
- mark know runners as deprecated

## Context
- closes #27290

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/github-runners/index.ts [hunk 1]
```diff
@@ -9,17 +9,25 @@ export class GithubRunnersDatasource extends Datasource {
    * Only add stable runners to the datasource. See datasource readme for details.
    */
   private static readonly releases: Record<string, Release[] | undefined> = {
-    ubuntu: [{ version: '22.04' }, { version: '20.04' }, { version: '18.04' }],
+    ubuntu: [
+      { version: '22.04' },
+      { version: '20.04' },
+      { version: '18.04', isDeprecated: true },
```
Comment (HonkingGoose) [inline lib/modules/datasource/github-runners/index.ts:15]:
```suggestion
      { version: '18.04', isDeprecated: true },
      { version: '16.04', isDeprecated: true },
```

Now we're deprecating old stuff, should we add and deprecate `16.04` as well?

- https://github.com/actions/runner-images/pull/8857

**Comment by viceice** [comment]
more tests are coming in #27292

File: lib/modules/datasource/github-runners/index.ts [hunk 2]
```diff
@@ -9,17 +9,26 @@ export class GithubRunnersDatasource extends Datasource {
...
+      { version: '16.04', isDeprecated: true },
+    ],
     macos: [
       { version: '13' },
       { version: '13-large' },
       { version: '13-xlarge' },
       { version: '12' },
       { version: '12-large' },
```
Comment (HonkingGoose) [inline lib/modules/datasource/github-runners/index.ts:23]: The upstream readme mentions `macos-latest`, `macos-12`, `macos-latest-large`, and `macos-12-large`. We don't have all those labels in our list yet.

https://github.com/actions/runner-images/blob/main/README.md#available-images

Comment (secustor) [reply lib/modules/datasource/github-runners/index.ts:23]: The `latest` are non-op for us as we have no pinning for these right now.    

Regarding tags `macos-` will be prefixed to the versions in the manager, if I recall correctly.

**Review by secustor — APPROVED** [review]

File: lib/modules/datasource/github-runners/index.ts (hunk 2 above)
Comment (HonkingGoose) [reply lib/modules/datasource/github-runners/index.ts:23]: Oh okay, just wanted to mention the apparently missing labels. 😄

**Review by HonkingGoose — APPROVED** [review]
It looks good to me, but I'll leave the final review to the maintainers.

**Pull request merged**